app/services/completions_decider.rb

```diff
@@ -0,0 +1,34 @@
+class CompletionsDecider
+  def initialize(user_agent:, request_url:)
+    @user_agent = user_agent
+    @request_url = request_url
+  end
+
+  def continue_to_sp?
+    user_agent_matches_redirect_uri?
+  end
+
+  private
+
+  attr_reader :user_agent, :request_url
+
+  def user_agent_matches_redirect_uri?
+    desktop_and_web_redirect_uri? || mobile_and_app_redirect_uri?
```

Comment:
I think we should definitely allow visiting the web app from mobile, our mail goal is disallowing native apps from desktop -- so maybe we just want to prevent that case rather than try to disallow the rest? (ex flip the logic of the controller?)